doc/modules/gaussian_process.rst

```diff
@@ -605,6 +605,100 @@ shown in the following figure:
    :target: ../auto_examples/gaussian_process/plot_gpr_prior_posterior.html
    :align: center
 
+Custom kernel
+------------------
+If you want to implement a new custom gaussian process kernel that is scikit-learn-compatible,
+you should subclass `sklearn.gaussian_process.kernels.Kernel` and implement
+the abstract methods `__call__`, `diag`, and `is_stationary`.
+
+Here is a step-by-step construction of the Sigmoid Kernel
+
+.. math::
+   k(x_i, x_j)= \tanh(\alpha x_i \cdot x_j + \sigma)
+
+
+First, we read in :math:`\alpha` and :math:`\sigma` and store these
+hyperparameters::
+
+   from sklearn.gaussian_process.kernels import Kernel, Hyperparameter
+
+   class MySigmoidKernel(Kernel):
+      def __init__(self, alpha_0=0.05,
+                  alpha_0_bounds=(1e-5, 1e5),
+                  sigma_0=5,
+                  sigma_0_bounds=(1e-5, 1e5)):
+         self.alpha_0 = alpha_0
+         self.alpha_0_bounds = alpha_0_bounds
+         self.sigma_0 = sigma_0
+         self.sigma_0_bounds = sigma_0_bounds
+
+      @property
+      def hyperparameter_scaling(self):
+         return Hyperparameter("alpha_0", "numeric", self.alpha_0_bounds)
+
+      @property
+      def hyperparameter_shifting(self):
+         return Hyperparameter("sigma_0", "numeric", self.sigma_0_bounds)
+
+Note that your custom kernel must be positive-definite for the choice of the
+hyperparameters.
+
+Now, from the kernel formula, we know that this kernel is non-stationary. If your custom
+kernel is stationary, you could subclass
+`sklearn.gaussian_process.kernels.StationaryKernelMixin` and skip this step; otherwise, we
+implement the method `is_stationary` as follows::
+
+      def is_stationary(self):
+         return False
+
+To calculate the diagonal, we will apply the kernel function on `X` (if your custom
+kernel is normalized, i.e. :math:`k(x,x)=1`, then you can subclass
+`sklearn.gaussian_process.kernels.NormalizedKernelMixin` and skip this step)::
+
+      def diag(self, X):
+         return np.apply_along_axis(self, 1, X).ravel()
```

Comment:
Using `__call__` would compute the whole matrix. Since the point of `diag` is be more efficient, is the following faster?

```py
np.tanh(self.alpha * np.einsum('ij,ij->i', X, X) + self.sigma)
```